cmd/clusternet-controller-manager/app/context/context.go

```diff
@@ -0,0 +1,94 @@
+package context
+
+import (
+	"context"
+
+	"github.com/clusternet/clusternet/cmd/clusternet-controller-manager/app/options"
+	clusternet "github.com/clusternet/clusternet/pkg/generated/clientset/versioned"
+	informers "github.com/clusternet/clusternet/pkg/generated/informers/externalversions"
+	"k8s.io/apimachinery/pkg/util/sets"
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/controller-manager/pkg/clientbuilder"
+	"k8s.io/klog/v2"
+	mcsclientset "sigs.k8s.io/mcs-api/pkg/client/clientset/versioned"
+	mcsinformers "sigs.k8s.io/mcs-api/pkg/client/informers/externalversions"
+)
+
+// ControllerContext defines the context object for controller.
+type ControllerContext struct {
+	Opts                      *options.ControllerManagerOptions
+	KubeConfig                *rest.Config
+	KubeClient                *kubernetes.Clientset
+	ElectionClient            *kubernetes.Clientset
+	ClusternetClient          *clusternet.Clientset
+	McsClient                 *mcsclientset.Clientset
+	ClusternetInformerFactory informers.SharedInformerFactory
+	KubeInformerFactory       kubeinformers.SharedInformerFactory
+	McsInformerFactory        mcsinformers.SharedInformerFactory
+	ClientBuilder             clientbuilder.ControllerClientBuilder
+	EventRecorder             record.EventRecorder
+
+	// InformersStarted is closed after all of the controllers have been initialized and are running.  After this point it is safe,
+	// for an individual controller to start the shared informers. Before it is closed, they should not.
+	InformersStarted chan struct{}
+}
+
+// StartControllers starts a set of controllers with a specified ControllerContext
+func (c ControllerContext) StartControllers(ctx context.Context, initializers Initializers) error {
```

Comment:
Shall we use the pointer receiver instead of struct?

Also `StartControllers` is only called once. So I'd prefer to have it as `StartControllers(ctx context.Context, controllerCtx ControllerContext, initializers Initializers)`.

Reply:
In the below `controllermanager.go`, a pointer is declared.

```go
controllerContext: &controllercontext.ControllerContext{...}
```

Please make them consistent.